apps/dapp/src/providers/SwapProvider.tsx

```diff
@@ -1,125 +1,98 @@
-import { useState, useContext, createContext, PropsWithChildren } from 'react';
-import { BigNumber, Signer, utils } from 'ethers';
+import { useState, useContext, createContext, PropsWithChildren, useEffect } from 'react';
+import { BigNumber, Contract, ethers } from 'ethers';
 import { TransactionReceipt } from '@ethersproject/abstract-provider';
 import { useWallet } from 'providers/WalletProvider';
 import { useNotification } from 'providers/NotificationProvider';
 import { SwapService } from 'providers/types';
-import { NoWalletAddressError } from 'providers/errors';
 import { TICKER_SYMBOL } from 'enums/ticker-symbol';
 import { asyncNoop } from 'utils/helpers';
-import { fromAtto } from 'utils/bigNumber';
-import { ERC20__factory, TempleERC20Token__factory, TempleUniswapV2Pair__factory } from 'types/typechain';
+import { ERC20__factory, TempleERC20Token__factory } from 'types/typechain';
 import env from 'constants/env';
 import { AnalyticsEvent } from 'constants/events';
 import { AnalyticsService } from 'services/AnalyticsService';
 import { formatBigNumber, getTokenInfo } from 'components/Vault/utils';
+import { SwapInfo } from '@balancer-labs/sor';
+import { BalancerSDK, Network, SwapType } from '@balancer-labs/sdk';
+import vaultArtifact from 'data/abis/balancerVault.json';
+import { formatToken } from 'utils/formatter';
+
+// Initialize balancer SOR
+const gasPrice = BigNumber.from('14000000000');
+const maxPools = 4;
+const balancer = new BalancerSDK({
+  network: Network.MAINNET,
+  rpcUrl: `https://eth-mainnet.g.alchemy.com/v2/AorwfDdHDsEjIX4HPwS70zkVjWqjv5vZ`,
+  // rpcUrl: `https://eth-mainnet.g.alchemy.com/v2/${env.alchemyId}`,
+});
+const sor = balancer.sor;
 
 const INITIAL_STATE: SwapService = {
-  templePrice: 0,
   buy: asyncNoop,
   sell: asyncNoop,
   getSellQuote: asyncNoop,
   getBuyQuote: asyncNoop,
-  updateTemplePrice: asyncNoop,
   error: null,
-  get1inchSwap: asyncNoop,
+  sor: balancer.sor,
+};
+
+// Compute slippage tolerance
+function getLimits(quote: SwapInfo, slippage: number): string[] {
+  const limits: string[] = [];
+  const slippageBps = BigNumber.from(10_000 + 100 * slippage);
+  quote.tokenAddresses.forEach((token, i) => {
+    // tokenIn = Max to send, should be positive
+    // tokenOut = Min to receive, should be negative
+    // Intermediate tokens should be 0 for multihops
+    if (token.toLowerCase() === quote.tokenIn.toLowerCase()) limits[i] = quote.swapAmount.toString();
+    else if (token.toLowerCase() === quote.tokenOut.toLowerCase())
+      limits[i] = quote.returnAmount.mul(10_000).div(slippageBps).mul(-1).toString();
```

Comment:
What happens if there are multiple hops? eg `limits[5]` could be the return token (in which case setting to `returnAmount` could be right), or it could be any of the other hops in which case it shouldn't be the `returnAmount`?

Reply:
this is how balancer does it on their frontend so I don't think it causes issues if you pass max/min limits on internal hops